crud/common/sharding.lua

```diff
@@ -20,14 +27,29 @@ function sharding.tuple_get_bucket_id(tuple, space, specified_bucket_id)
         return specified_bucket_id
     end
 
-    local key = utils.extract_key(tuple, space.index[0].parts)
+    local primary_index = space.index[0]
+    local key
+    local replicasets = vshard.router.routeall()
+    local replicaset = select(2, next(replicasets))
+    local sharding_key = sharding.get_ddl_sharding_key(space.name, replicaset.master.conn)
+    if sharding_key ~= nil then
+        local sharding_keys_fieldnos = utils.get_keys_fieldnos(space:format(), sharding_key)
+        if sharding_keys_fieldnos ~= nil then
+            key = utils.extract_sharding_key(tuple, sharding_keys_fieldnos)
+        end
+    end
+
+    if key == nil then
+        key = utils.extract_key(tuple, primary_index.parts)
+    end
+
     return sharding.key_get_bucket_id(key)
 end
 
 function sharding.tuple_set_and_return_bucket_id(tuple, space, specified_bucket_id)
     local bucket_id_fieldno, err = utils.get_bucket_id_fieldno(space)
     if err ~= nil then
-        return nil, BucketIDError:new("Failed to get bucket ID fielno: %s", err)
+        return nil, BucketIDError:new("Failed to get bucket ID fieldno: %s", err)
```

Comment:
Let's move such typos to separate PR and I'm ready to approve it instantly.

Reply:
I will do that once we decide that review has passed.

Reply:
https://github.com/tarantool/crud/pull/194